src/scheduled_tasks.rs

```diff
@@ -70,66 +70,127 @@ pub fn setup(db_url: String, user_agent: String) -> Result<(), LemmyError> {
 }
 
 /// Update the hot_rank columns for the aggregates tables
+/// Runs in batches until all necessary rows are updated once
 fn update_hot_ranks(conn: &mut PgConnection, last_week_only: bool) {
-  let mut post_update = diesel::update(post_aggregates::table).into_boxed();
-  let mut comment_update = diesel::update(comment_aggregates::table).into_boxed();
-  let mut community_update = diesel::update(community_aggregates::table).into_boxed();
+  let update_start_time = naive_now();
+  let update_batch_size = 5000;
+  let last_week = now - diesel::dsl::IntervalDsl::weeks(1);
 
-  // Only update for the last week of content
   if last_week_only {
     info!("Updating hot ranks for last week...");
-    let last_week = now - diesel::dsl::IntervalDsl::weeks(1);
-
-    post_update = post_update.filter(post_aggregates::published.gt(last_week));
-    comment_update = comment_update.filter(comment_aggregates::published.gt(last_week));
-    community_update = community_update.filter(community_aggregates::published.gt(last_week));
   } else {
     info!("Updating hot ranks for all history...");
   }
 
-  match post_update
-    .set((
-      post_aggregates::hot_rank.eq(hot_rank(post_aggregates::score, post_aggregates::published)),
-      post_aggregates::hot_rank_active.eq(hot_rank(
-        post_aggregates::score,
-        post_aggregates::newest_comment_time_necro,
-      )),
-    ))
-    .execute(conn)
-  {
-    Ok(_) => {}
-    Err(e) => {
-      error!("Failed to update post_aggregates hot_ranks: {}", e)
+  loop {
+    let mut post_select = post_aggregates::table
+      .select(post_aggregates::post_id)
+      .filter(post_aggregates::hot_rank_updated.lt(update_start_time))
```

Comment:
Hmm, but in this case, if there is some hugely popular post (or even comment thread) within some time bucket, then the load will be spread quite unevenly and we can still end up with an UPDATE statement that can have several tens (or why not even hundreds) of thousands of touched rows. Or am I misunderstanding you?

Reply:
I'd say even with that, it'd still be *good enough* for batching, especially since that number is tunable and simple. As long as there's enough buckets we should be okay. IE 7, 1-day buckets might even be enough to prevent the deadlocks, but in code we could make that `1000` buckets.

The other alternative is doing a select to gather up all the IDs for the last 7 days, then doing an `eq_any` / in query, although those are generally extremely inefficient and wasteful.

Reply:
You could also do it without bucketing by doing "normal" timestamp pagination: (pseudocode)

last_timestamp = '1970-01-01' // or last week i guess
do {
rows = select comment_id from comments where published > last_timestamp order by published asc limit batch_size;
last_timestamp = rows[-1].published;
rows = rows.sortby(comment_id) // needs to be consistent with other multi-row update queries for deadlocks
} while len(rows) == batch_size

* batches have fixed sizes
* no parallelization possible because next batch is only determined after current batch
* need to look for off-by-one error if two rows have the same timestamp


advantage of the last_update column is that you know which rows still need updating but yeah seems like it might not be needed at this point

Reply:
Yeah, I'm not sure why I didn't go with that approach to begin with - I remember considering it and then thinking it would potentially never finish processing in some cases, but I think I was just a bit sleep deprived and confused myself 😄 

@dessalines would you be OK with that approach as well?

Reply:
I'm not exactly sure how that would translate to SQL / diesel, and it seems a bit more complicated than just creating a vector of equally-spaced times with a *between* query, but sure. As long as it doesn't use an `eq_any` / in query, it should be fine.